invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
+func (e invoiceExpiry) Less(other queue.PriorityQueueItem) bool {
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
+// invoices.
+// Upon start InvoiceExpiryWatcher will subscribe to all invoice events and
+// will add all active (not yet settled and not yet cancelled) invoices to its
+// watcing queue.
+// When a new invoice is added to the InvoiceRegistry, InvoiceExpiryWatcher
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
+	sync.Mutex
+	started bool
+	wg      sync.WaitGroup
+
+	// clock is the clock implementation that InvoiceExpiryWatcher uses.
+	// It is useful for testing.
+	clock clock.Clock
+
+	// quit signals InvoiceExpiryWatcher to stop.
+	quit chan interface{}
+
+	// cancelInvoice is a template method that cancels an expired invoice.
+	cancelInvoice func(lntypes.Hash) error
+
+	// expiryQueue holds invoiceExpiry items and is used to find the next
+	// invoice to expire.
+	expiryQueue queue.PriorityQueue
+
+	// newInvoices channel is used to wake up the main loop when a new invoices
+	// is added.
+	newInvoices chan *invoiceExpiry
+}
+
+// NewInvoiceExpiryWatcher creates a new InvoiceExpiryWatcher instance.
+func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+	return &InvoiceExpiryWatcher{
+		clock:       clock,
+		quit:        make(chan interface{}),
+		newInvoices: make(chan *invoiceExpiry),
+	}
+}
+
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher. This is useful to
+// prepopulate the watcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) {
+	pendingOnly := true
+	invoices, err := cdb.FetchAllInvoices(pendingOnly)
+	if err != nil {
+		log.Errorf("Error fetching invoices from the database: %v", err)
+	} else {
+		for k := 0; k < len(invoices); k++ {
+			var paymentHash lntypes.Hash
+			copy(paymentHash[:lntypes.HashSize], invoices[k].PaymentHash)
+			ew.AddInvoice(paymentHash, &invoices[k])
+		}
+	}
+}
+
+// Start starts the the subscription handler and the main loop. Start() can
+// only be called once.
+func (ew *InvoiceExpiryWatcher) Start(
+	cancelInvoice func(lntypes.Hash) error) error {
+	ew.Lock()
+	defer ew.Unlock()
+
+	if ew.started {
+		return fmt.Errorf("InvoiceExpiryWatcher already started")
+	}
+
+	ew.started = true
+	ew.cancelInvoice = cancelInvoice
+	ew.wg.Add(1)
+	go ew.mainLoop()
+
+	return nil
+}
+
+// Stop cancels the invoice subscription and stops the expiry handler loop.
+func (ew *InvoiceExpiryWatcher) Stop() {
+	ew.Lock()
```

Comment:
Haven't seen an instance in this diff, but my spidey senses tell me that a deadlock may be inadvertently introduced here if all callers don't mind proper mutex usage. We use atomic variables in every other new sub-system within the codebase to side step this, as things quickly exit if a shutdown has already been initiated.

Reply:
The lock is only locking the Start/Stop procedure. Such that `Stop` will wait for `Start` if signaled during `Start`. Could you elaborate on how this may deadlock?

Reply:
I think if we'd use the lock inside the main loop as well. Agreed that this looks a bit risky. It may be correct now, but a future dev may use this lock for something else too.